## 错误监控

Error对象上主要有name、message、stack这三个有用的属性。
- name：错误的类型
- message：错误提示信息
- stack：错误栈
我们可以使用Error.prototype.toString快速获取name+message的错误信息字符串。

### try...catch
```
try {
  throw new Error('test');
} catch (error) {
  handleError(error); // handleError是自定义错误信息处理函数
}
```
### global.ErrorUtils
```
// 存放原始的错误处理函数
const errorHandle = global.ErrorUtils.getGlobalHandler();

global.ErrorUtils.setGlobalHandler(handleError, true); // handleError是自定义错误信息处理函数
```

### handleError
```
function lsErrorStack(e) {
  if (!e || !e instanceof Error || !e.stack) return {};

  try {
    const stack = e.stack.toString().split(/\r\n|\n/), frameRE = /:(\d+:\d+)[^\d]*$/;
    
    while (stack.length) {
      const frame = frameRE.exec(stack.shift())
      if (frame) {
        const position = frame[1].split(':')
        return { line: position[0], column: position[1] };
      }
    }
  } catch () {
    return {}
  }
}

function handleError(e, isFatal) {
  const position = lsErrorStack(e);
  
  // 上报错误信息：e.toString()； position；等信息
  report(...)
}
```

### 未捕获的promise错误
RN源码中就有监控未捕获的promise错误的代码，我们可以将代码ctl+c、ctl+v，并注意在非开发环境下监控
```
if (!__DEV__) {
  require('promise/setimmediate/rejection-tracking').enable({
    allRejections: true,
    onUnhandled: (id, error) => {
      // 错误信息上报
    },
    onHandled: (id) => {},
  });
}
```





